plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/ResourceContainer.java

```diff
@@ -0,0 +1,30 @@
+package org.jboss.tools.openshift.internal.ui.models;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.openshift.restclient.model.IResource;
+
+public abstract class ResourceContainer<R extends IResource, P extends IOpenshiftUIElement<?>>
+		extends AbstractResourceWrapper<R, P> {
+	protected Map<IResource, AbstractResourceWrapper<?, ?>> containedResources = new HashMap<>();
+	protected Object childrenLock = new Object();
+
+	public ResourceContainer(P parent, R resource) {
+		super(parent, resource);
+	}
+
+	public Collection<AbstractResourceWrapper<?, ?>> getResourcesOfKind(String kind) {

```

Comment:
The model will not refresh anything, unless the the user explicitly ask to do so. Otherwise, we are getting change notifications for single resources. Since we show resources as contained inside Services, a change of a label, for example might move a number of resources to another service. Yes, recalculating the containement hierarchy on each resource change may be expensive, but it will only touch stuff that is in-memory and should be reasonably fast. Also, the changes are merged in the background and only notifications about changes in UI model are notified on the UI thread.